plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/ASTPreProcess.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ide.Plugin.SelectionRange.ASTPreProcess
+    ( preProcessAST
+    , PreProcessEnv(..)
+    ) where
+
+import           Control.Monad.Reader            (Reader, asks)
+import           Data.Foldable                   (find, foldl')
+import           Data.List                       (groupBy)
+import qualified Data.Map.Strict                 as Map
+import           Data.Maybe                      (mapMaybe)
+import qualified Data.Set                        as Set
+import           Development.IDE.GHC.Compat      (ContextInfo (MatchBind, TyDecl, ValBind),
+                                                  HieAST (..), Identifier,
+                                                  IdentifierDetails (identInfo),
+                                                  NodeInfo (NodeInfo, nodeIdentifiers),
+                                                  RefMap, Span, flattenAst,
+                                                  isAnnotationInNodeInfo,
+                                                  mkAstNode, mkRealSrcSpan,
+                                                  nodeInfoFromSource,
+                                                  realSrcSpanEnd,
+                                                  realSrcSpanStart)
+import           Development.IDE.GHC.Compat.Util (FastString)
+import           Prelude                         hiding (span)
+
+newtype PreProcessEnv a = PreProcessEnv
+    { preProcessEnvRefMap :: RefMap a
+    }
+
+-- | Make the AST more suitable for generating selection range.
```

Comment:
Could briefly explain the changes and why we need them?

Reply:
This is used to handle some cases not covered by the default tree structure of `HieAST`. Will supply more detailed explaination in comment.